src/core/service.c

```diff
@@ -3549,14 +3553,42 @@ static int service_dispatch_exec_io(sd_event_source *source, int fd, uint32_t ev
                                 log_unit_debug(UNIT(s), "Got EOF on exec-fd while it was disabled, ignoring.");
 
                         return 0;
+
+                case sizeof(uint8_t):
+                        /* A single byte was read, which means the exec fd logic is now off (unless somehow
+                         * an old executor has written back to a new pid1, so also allow x=1 just in case).
+                         * Also reset the exec timestamp, given the execve actually failed. */
+
+                        s->exec_fd_hot = x;
+                        if (s->state == SERVICE_START)
+                                s->main_exec_status.handover_timestamp = DUAL_TIMESTAMP_NULL;
+
+                        continue;
+
+                case sizeof(uint64_t): {
+                        /* A timestamp was read, which means the exec fd logic is now enabled. Record the
+                         * received timestamp so that users can track when control is handed over to the
+                         * service payload. */
+                        assert_cc(sizeof(uint64_t) == sizeof(usec_t));
+                        usec_t t = x;
+
+                        s->exec_fd_hot = true;
+                        if (s->state == SERVICE_START)
+                                dual_timestamp_from_monotonic(&s->main_exec_status.handover_timestamp, t);
+
+                        continue;
                 }
 
-                /* A byte was read → this turns on/off the exec fd logic */
-                assert(n == sizeof(x));
-                s->exec_fd_hot = x;
+                default:
+                        assert(n < 0);
+                        if (errno == EAGAIN) /* O_NONBLOCK in effect → everything queued has now been processed. */
+                                return 0;
+
+                        return log_unit_error_errno(UNIT(s), errno, "Failed to read from exec_fd: %m");
+                }
         }
 
-        return 0;
+        assert_not_reached();
```

Comment:
Is this line necessary? The compiler usually can figure out that a line is unreachable after such a loop.